Версия 3.4

Mazdaywik opened the pull request:
В версию 3.4 войдут, прежде всего расширенные прогонка и специализация, написанные @Apakhov’ым и @VladisP’ом, этот код уже в `master`.

Данный pull request вливает:
* #359 — доработки расширенной прогонки и специализации, позволяющие самоприменить компилятор с ключами `-OiADPRS`).

Mazdaywik committed:
FIXED: проблема со специализацией замыканий (#251, #359)

Ранее специализация замыкания {{ &F CONTENT }} осуществлялась как
специализация фиктивного вызова <F CONTENT e.@>. Оптимизатор строил новый
вызов <F@1 CONTENT′ e.@>, из которого восстанавливалось замыкание
{{ &F@1 CONTENT′ }}.

Добавлять и удалять фиктивную переменную e.@ было безопасно, т.к. она была
в позиции динамического параметра, позиция параметра не менялась (оставалась
последней) и во внутрь функции она не протекала.

Теперь же при специализации это имя может протечь внутрь экземпляра,
внутри экземпляра может оказаться другое замыкание, в контексте которого
будет переменная e.@. Добавление e.@ в конец вызовет конфликт имён.

----

В коде есть некоторый костыль, который проистекает из отсутствия поддержки
подавления предупреждений. Исходники намеренно собираются с -Werror,
а в данном коде было бы разумно явно добавить «подозрительную» (#290)
повторную переменную. Пришлось достигать эту проверку более громоздким
куском исходного кода.

Mazdaywik committed:
Повышение предела числа шагов для самоприменения компилятора (#359)

Самоприменение выполняется в режиме

set RLMAKE_FLAGS=-X-OiDPRS

Mazdaywik committed:
FIXED: прогонка функций с пустым телом приводила к падению (#359)

Ошибка была внесена в коммите 204265358888 в рамках задачи #303.

Mazdaywik committed:
Рефакторинг: удалены неиспользуемые функции (#359)

Mazdaywik committed:
Оптимизация: переписана проверка отношения Хигмана-Крускала (#359)

До оптимизации после самоприменения в режиме:

set RLMAKE_FLAGS=-X-OiDPRS
call makeself.bat
set RLMAKE_FLAGS=-X-OiDS
call makeself.bat

со включённым профилем время работы составило 1590 секунд, где-то 50 %
времени ушло на проверку отношения Хигмана-Крускала.

В актуальной реализации время работы составило 575 секунд, отношение
Хигмана-Крускала занимает где-то около процента.

Подробности с логами профилировщика в issue.

Mazdaywik committed:
Микрооптимизация GrowDriveTree (#359)

Mazdaywik committed:
Рефакторинги: разбиения сложных функций на несколько простых (#359)

Рекомендуется смотреть коммит с игнорированием пробельных символов.

Mazdaywik committed:
Стилевые правки: форматирование кода (#359)

Mazdaywik committed:
Рефакторинг: уточнение имени функции, лишние комментарии (#359)

Mazdaywik committed:
Оптимизация MakeDriveTree/GrowDriveTree (#359)

Mazdaywik committed:
FIXED: при сборке исходников отключён встроенный профилировщик (#359)

Эта правка была закоммичена в предыдущем коммите по ошибке.

Mazdaywik committed:
Мелкие рефакторинги, стилевые правки (#322, #359)

• Переименования функций,
• уточнения комментариев,
• изменение типа результата функции Solve-Clashes,
• удалена неиспользуемая функция IsFreeVariableSeq.

Mazdaywik committed:
Сохранение используемых переменных при прогонке (#359)

Ранее при построении дерева прогонки, из сужений для ветвей извлекались
имена используемых переменных. Теперь эти имена возвращаются функцией
Solve-Drive. На быстродействие это почти не повлияло.

Ранее:

ExtractVariables-Expr (0000) -> 11519.0 ms (2.64 %, += 21.06 %)
                       rel step time 0.67
ExtractVariables-Expr (0000) -> 13593437 (3.92 %, += 8.41 %)
                       rel step time 0.67

Теперь:

ExtractVariables-Expr (0000) -> 11687.0 ms (2.73 %, += 21.22 %)
                       rel step time 0.71
ExtractVariables-Expr (0000) -> 13442507 (3.87 %, += 8.34 %)
                       rel step time 0.71

Число шагов уменьшилось на 0,05 % (3,92−3,87), а доля времени даже
выросла.

Mazdaywik committed:
Уточнение комментариев (#322, #359)

Mazdaywik committed:
Удалена функция SimplifyCoordinates-Expr-Inner (#359)

Данная правка на время и число шагов функции SimplifyCoordinates-Expr,
вопреки ожиданиям, никак не повлияла.

Ранее:

Total time: 427.50000 s

SimplifyCoordinates-Expr (3646) -> 6957.0 ms (1.63 %, += 26.16 %)
                    rel step time 0.90
SimplifyCoordinates-Expr (3646) -> 6259417 (1.80 %, += 24.36 %)
                    rel step time 0.90

SimplifyCoordinates-Expr-Inner (3646) -> 6274.0 ms (1.47 %, += 29.18 %)
              rel step time 1.17
SimplifyCoordinates-Expr-Inner (3646) -> 4372638 (1.26 %, += 30.18 %)
              rel step time 1.17

Сейчас:

Total time: 422.06200 s

SimplifyCoordinates-Expr (5246) -> 6909.0 ms (1.64 %, += 28.27 %)
                    rel step time 0.90
SimplifyCoordinates-Expr (5246) -> 6222848 (1.81 %, += 24.66 %)
                    rel step time 0.90

Число шагов SimplifyCoordinates-Expr даже, почему-то, возросло.

Но SimplifyCoordinates-Expr-Inner больше нет, программа ускорилась
на сэкономленные ≈6 секунд. Понятно, что это однократный замер,
просто интересно совпало.

Mazdaywik committed:
Встраивание условий-сужений в Solve-Clashes (#322, #359)

Mazdaywik committed:
Оптимизация отделения термов (#322, #359)

Mazdaywik committed:
FIXED: зацикливание клэша с замыканием слева (#322, #359)

Mazdaywik committed:
Лимит числа шагов для обычных автотестов снижен до 60 000 000 (#359)

Этот лимит инкрементировался в рамках задач #251, #322, #340.

Mazdaywik committed:
FIXED: внутри конструкторов замыканий не применялись сужения (#322, #359)

Mazdaywik committed:
Стилевая правка: правильное использование ADT-Brackets (#359)

Mazdaywik committed:
Мелкие стилевые правки в GenericMatch.ref (#359)

Mazdaywik committed:
Уточнено определение тривиальной сигнатуры (#251, #359)

Mazdaywik committed:
Рефакторинг: устранено дублирование кода (#359)

Mazdaywik committed:
$EXTERN’ы отсортированы в порядке использования (#359)

Mazdaywik committed:
Рефакторинг: ExpandClosures-Result перенесена в TreeUtils.ref (#359)

Mazdaywik committed:
FIXED: уточнена классификация выражений (#359)

В коммите 286a098a9 классификация была неточная,
но к ошибкам это не приводило.

Mazdaywik committed:
Рефакторинг проверки на пассивное выражение (#359)

Mazdaywik committed:
Небольшие рефакторинги в OptTree-Drive-Expr.ref (#359)

Mazdaywik committed:
Рефакторинг: выделение OptTree-Drive-Intrinsic в отдельный файл, шаг 1 (#359)

Файл OptTree-Drive-Expr.ref переименован для сохранения истории правок.

Mazdaywik committed:
Рефакторинг: выделение OptTree-Drive-Intrinsic в отдельный файл, шаг 2 (#359)

Техническое слияние двух веток. Неиспользуемые функции в файлах
OptTree-Drive-Expr.ref и OptTree-Drive-Intrinsic.ref удалены.
История правок, отображаемая в git blame, сохранена.

Mazdaywik committed:
Рефакторинг: прояснение построения дерева прогонки (#340, #359)

Mazdaywik committed:
Рефакторинг: учёт того, что сужения применяются последовательно (#359)

Mazdaywik committed:
Рефакторинг функции FlatDriveTree (#359)

Mazdaywik committed:
FIXED: развёртка вызовов замыканий при оптимизациях (#251, #340, #359)

Mazdaywik committed:
FIXED: исправление двух ошибок (#322, #359)

• Первая ошибка в том, что функция расстановки координат не учитывала
  холодные скобки вызова, из-за чего координаты размечала неправильно.
  (Неправильно она размечала координаты из-за неправильного выхода
  из рекурсии).
• Вторая ошибка, которая выявила первую ошибку, была допущена в коммите
  add183e6. В функции определения типов выражений для некоторых аргументов
  открытые переменные имели приоритет над наличием скобок вызова.

Mazdaywik committed:
Мелкий рефакторинг ExprType-Rec (#359)

Mazdaywik committed:
Рефакторинг: упрощён и документирован тип OptTree-Drive-Intrinsic (#359)

Mazdaywik committed:
Избыточные копирования переменных, пропущенные #290 (#359)

Формально, данные повторные переменные не подходят под критерий #290:
однозначное вхождение в последнем предложении. Но в случае второго
предложения здесь:

    : {
      Generalize t.OptInfo^ 1 = …
      Generalize t.OptInfo^ s.Cnt = …
      Tree t.DriveTreeState e.Branches^ = …
    }

можно выдать предупреждение на t.OptInfo по следующему критерию —
последующие предложения не способны поглотить аргумент, совпадающий
с данным. И поэтому переменная подозрительная.

Так что, возможно, критерий стоит уточнить.

Mazdaywik committed:
Оптимизации копирования переменных для режима -OAiDS (#359)

Mazdaywik committed:
Мелкий рефакторинг к предыдущему коммиту 49c50613 (#359)

Mazdaywik committed:
Предварительная проверка по отношению Хигмана в условии остановки (#359)

Подавляющее большинство проверок на отношение Хигмана-Крускала
даёт отрицательный ответ, поэтому имеет смысл предварительно проверять
более быстрое отношение Хигмана, являющее необходимым условием
для отншения Хигмана-Крускала.

В коммит включена возможность включить отладочный код, выводящий
статистику успешности отношений Х. и Х.-К.

Mazdaywik committed:
Оптимизации копирования переменных для режима -OAiDS (#359)

Mazdaywik committed:
Специализация вызовов и замыканий снаружи вовнутрь (#359)

Во-первых, исправляется проблема со специализацией замыканий вида

    {{ &F\1 … {{ &G\1 … }} … }}

когда наружнее &F\1 специализируется по специализированному внутреннему &G\1:

    {{ &F\1 … {{ &G\1 … }} … }}
                 ↓
    {{ &F\1 … {{ &G\1@1 … }} … }}
                 ↓
    {{ &F\1@1 … {{ &G\1@1 … }} … }}

Во-вторых, устраняются два прохода: сначала функции, потом замыкания. Два
прохода, фактически, были костылём.

Mazdaywik committed:
Рефакторинг: устранение артефактов двухпроходной специализации (#359)

Mazdaywik committed:
$INLINE Map, Reduce, DoMapAccum (#359)

Mazdaywik committed:
Принудительное обобщение аккумулятора в DoUnEscapeString-SR (#359)

Mazdaywik committed:
Подавление древесной оптимизации в rl-lexgen при помощи gen_e__ (#359)

Ранее древенсая оптимизация в lexgen’е подавлялась при помощи объявления
функций как

    $SPEC State (e.acc) e.text;

До реализации #251 «Специализация без шаблона» эта строчка подавляла
специализацию функции-состояния в режиме -OA+. Сейчас эта строчка наоборот,
форсирует специализацию этой функции даже в режиме -OA-.

Избыточная прогонка подавлена при помощи организации автомата так, чтобы
авторазметка (#251) рассматривала некоторые «опасные» функции как базисные,
см. 0cda99e2d.

В текущей версии 3.3 реализована поддержка функции обобщения gen_e__ (#331),
которая предназначена для подавления специализации. Теперь все аккумуляторы
принудительно обобщаются при помощи gen_e__. Прогонка подавляется за счёт
того, что в актуальной версии не реализована прогонка функций с активным
аргументом (#230).

Потеря быстродействия не страшна, т.к. Простой Рефал уже deprecated (#318,
DSL, который растворяется во время компиляции (#50).

Потеря быстродей

Mazdaywik committed:
Принудительные обобщения в куче мест для борьбы с распуханием (#359)

По-нормальному, нужно каждый случай обобщения описать, сделать выводы
и решить проблему в общем случае в рамках #332. А пока я анализировал
лог и подсекал наиболее явные точки распухания.

Mazdaywik committed:
$INLINE ReportErrors, ReportWarnings, ReportMessages; (#359)

ReportWarnings полностью встраивается, число экземпляров ReportErrors
существенно уменьшилось. Оставшиеся случаи — предложения с условиями
и не-L-образцами. И это тоже странно, т.к. там она должна встроиться.

Mazdaywik committed:
Принудительные обобщения в куче мест для борьбы с распуханием № 2 (#359)

По-нормальному, нужно каждый случай обобщения описать, сделать выводы
и решить проблему в общем случае в рамках #332. А пока я анализировал
лог и подсекал наиболее явные точки распухания.

Mazdaywik committed:
Принудительные обобщения в куче мест для борьбы с распуханием № 3 (#359)

По-нормальному, нужно каждый случай обобщения описать, сделать выводы
и решить проблему в общем случае в рамках #332. А пока я анализировал
лог и подсекал наиболее явные точки распухания.

Mazdaywik committed:
Рефакторинг: замена TermCompare → Compare (#359)

Mazdaywik committed:
FIXED: исправлено точное определение тривиальной сигнатуры (#251, #359)

Проблема была внесена коммитом c1de4f56. Ошибка заключалась в том, что
вызов функции IsTrivialSignature не соответствовал формату, поэтому
условие всегда было ложным.

Mazdaywik committed:
Стилевая правка: длинные строки в R5-Parser.ref (#359)

Mazdaywik committed:
Обобщение участков аргумента между двумя e-параметрами (#332, #359)

Mazdaywik committed:
Мелкие стилевые правки в OptTree-Spec.ref (#359)

Mazdaywik committed:
Принудительные обобщения в куче мест для борьбы с распуханием № 4 (#359)

По-нормальному, нужно каждый случай обобщения описать, сделать выводы
и решить проблему в общем случае в рамках #332. А пока я анализировал
лог и подсекал наиболее явные точки распухания.

В Log-AST.ref был интересный случай квадратичного распухания,
а в остальных местах — типичная специализация по аккумулятору.

Mazdaywik committed:
FIXED: возможность сборки стабильной версией (#359, #361)

Mazdaywik committed:
Проверка самоприменения в режиме -OiADPRS в CI (#323)

Mazdaywik committed:
Удалено сохранение ненужных артефактов в CI (#323)

Mazdaywik committed:
Упавшие случайные тесты теперь сохраняются как артефакты в CI (#323)

Mazdaywik committed:
FIXED: возможность сборки стабильной версией (#359, #361)

Mazdaywik committed:
Предупреждение GCC: учёт возвращаемого значения fgets() (#85)

Mazdaywik committed:
Рефакторинг: перенос ApplyContractions в TreeUtils.ref (#359)

Mazdaywik committed:
FIXED: проверка самоприменения в режиме -OiADPRS в CI (#323)

Повторная самосборка для Windows по ошибке включала автотесты.

Mazdaywik committed:
Небольшой рефакторинг упрощения координат (#359)

Mazdaywik committed:
FIXED: ошибка в документирующем комментарии (#359)

Mazdaywik committed:
Небольшой рефакторинг построения предложений экземпляра (#359)

Mazdaywik committed:
Сужения в решениях уравнений теперь последовательные (#359)

Mazdaywik committed:
Рефакторинг: перенос ApplyAssigns в TreeUtils.ref (#359)

Mazdaywik committed:
Использование последовательных сужений при специализации (#259)

Mazdaywik committed:
Почти ленивые генерация комментариев и логирование (#359)

Теперь для CmdComment и вывода сигнатур текстовые представления выражений
печатаются только в момент реального вывода. Для этого вместо самих
текстовых представлений передаются функции, их возвращающие.

«Почти ленивые», т.к. результат не мемоизируется, а это значит,
что при некоторых условиях текстовые представления выражений могут вычисляться
дважды (а именно, включён вывод в лог и -Od).

Mazdaywik committed:
Повышен лимит памяти для самоприменения комплятора до 100 000 000 (#359)

Mazdaywik committed:
Опечатки и ошибки в src/compiler/README.md (#149)

Mazdaywik committed:
Мелкий рефакторинг в back-end’е

Mazdaywik committed:
Алгоритм обобщ. сопоставления для проверки экранирования (#256, #359)

Старый код в ObjectMatch.ref уже не используется, он будет удалён
в следующем коммите для более ясной истории.

Новый алгоритм выявляет большее количество предупреждений, чем старый.
Автотесты демонстрируют такие случаи.

Mazdaywik committed:
Рефакторинг: удалён старый алгоритм в ObjectMatch.ref (#256, #359)

Mazdaywik committed:
Мелкие уточнения комментариев (#359)

Mazdaywik committed:
Рефакторинг: ObjectMatch перенесена в Checker-Screening.ref (#256, #359)

Mazdaywik committed:
Рефакторинг: устранён костыль с квадратными скобками (#256, #359)

Mazdaywik committed:
Упрощение проверки вложения языков образцов (#256, #359)

Выделять кортежи и сопоставлять отдельно элементы кортежей нет
необходимости. Если два выражения являются уточнением одного жётского
выражения, то при обобщённом сопоставлении с образцом их обобщение
выделится сразу несколькими «транзитными» шагами.

Mazdaywik committed:
Рефакторинг: удалены неиспользуемые функции (#256, #359)

Удаление сделано отдельным коммитом ради упрощения чтения истории.

Mazdaywik committed:
Мелкие стилевые правки в Checker-Screening.ref (#256, #359)

Mazdaywik committed:
Рефакторинг: удалено бесполезное предложение (#256, #369)

Mazdaywik committed:
Некоторый рефакторинг в Checker-Screening.ref (#256, #359)

Mazdaywik committed:
Стилевая правка: длинная строка в Checker-Screening.ref (#256, #359)

Mazdaywik committed:
Миниоптимизация функции Solve-SymClashes-Aux (#359)

Вместо трёх разных предложений, рассматривающих случай

    перем = перем

используется одно предложение

Mazdaywik committed:
FIXED: исправлено зацикливание в Solve-SymClashes-Aux (#256, #359)

Ошибка была в том, что для клэшей вида (ScrVar …) = ‹один-терм› не было
ни одного применимого правила, кроме правила отщепления термов. Из-за этого
единственный терм отщеплялся в цикле вечно.

Mazdaywik committed:
Мелкие стилевые правки в Desugaring.ref (#315, #359)

Mazdaywik committed:
Рефакторинг: удалено устаревшее предложение (#315, #359)

Mazdaywik committed:
(Swap s.Scope e.Name) → (Function s.Scope (e.Name) Swap) (#315)

Mazdaywik committed:
Стилевая правка: длинная строка в Engine.ref (#359)

Mazdaywik committed:
Мелкие стилевые правки в LowLevelRASL*.ref (#359)

Mazdaywik committed:
Узел дерева (Entries …), удаление s.ScopeClass из остальных узлов (#315)

Рефакторингом это назвать нельзя, т.к. как минимум изменился подсчёт
хешей единиц трансляции. Также, возможно, есть другие нейтральные
изменения в сгенерированном коде (не проверял).

Mazdaywik committed:
Разные стилевые правки (#315, #359)

• Исправлены нечётные отступы.
• Убраны пробелы в концах строк в недавно менявшихся строчках
  (или рядом с ними).
• Удалены лишние ссылки в списках $EXTERN.
• $EXTERN’ы упорядочены в порядке использования.

Коммит рекомендуется смотреть также с опцией игнорирования пробельных
символов.

Mazdaywik commented:
То, что упал коммит 43016ec47a0b5898d9d05e2cf9760b0c9c932b8e (`Мелкие стилевые правки в LowLevelRASL*.ref (#359)`), я заметил не сразу. Ошибка там есть, и она в процессе исправления.

А вот коммит 3d41447b2904e18ab6f9832107920295794162e0 (`Разные стилевые правки (#315, #359)`) упал каким-то незнакомым мне образом:

![image](https://user-images.githubusercontent.com/1309859/128666226-46e7b01c-36f8-4d9f-b784-856764dd18dc.png)

Mazdaywik committed:
FIXED: отделения термов теперь не нарушают инвариант (#322, #359)

В заявке #318 рассматривается возможность удаления рассахаривателя
условий -OC-. А ведь эта ошибка была найдена именно при компиляции
рассахаренного кода.

Mazdaywik commented:
Прервал тест «CI-posix / bootstap-sh (windows-latest, gcc)». В логе обнаружил такую картину:
```
2021-08-09T08:38:00.8973121Z Passing test-1_Mon-Aug--9-08-36-56-CUT-2021 (-OiCD):
2021-08-09T08:38:01.3645849Z Passing test-1_Mon-Aug--9-08-36-56-CUT-2021 (-OiACD):
2021-08-09T08:38:01.9333185Z Passing test-1_Mon-Aug--9-08-36-56-CUT-2021 (-OAS):
2021-08-09T08:38:03.7737278Z Passing test-1_Mon-Aug--9-08-36-56-CUT-2021 (-OADS):
2021-08-09T08:46:07.0764842Z ##[warning]You are running out of disk space. The runner will stop working when the machine runs out of disk space. Free space left: 0 MB
2021-08-09T08:56:22.1666316Z ##[error]The operation was canceled.
2021-08-09T08:56:22.3427317Z Post job cleanup.
```
Сообщает о том, что место на диске кончилось. Возможно, именно из-за этого предыдущий тест также странно упал (по таймауту).

Интересно, как такое могло получиться?

Mazdaywik committed:
Проверка свободного места на сервере CI (#361)

Mazdaywik commented:
А сейчас тесты прошли.

Можно предположить, что компилятор падал, порождая огромный дамп, который съедал всё доступное пространство. Сейчас тест, выявляющий эту ошибку, не сгенерировался и проблема не воспроизвелась.

Стоит добавить вывод размеров исходных файлов и свободного места в вывод случайных тестов. В этом случае можно будет визуально оценивать, какие тесты большие, а какие маленькие (приятно такое знать, когда наблюдаешь за процессом), а кроме того, будет понятно свободное место.

Mazdaywik committed:
Размеры файлов и свободного пространства для случайных тестов (#361)

На Windows команда dir также выводит объём свободного места.

Mazdaywik commented:
Места на серверах более чем достаточно:
* На Ubuntu — 38 Гбайт:
  ```
  Filesystem      Size  Used Avail Use% Mounted on
  /dev/root        84G   47G   38G  56% /
  ```
* На Windows — примерно 13 Гбайт:
  ```
                 10 File(s)         64,991 bytes
                  0 Dir(s)  12,836,265,984 bytes free
  ```
* На macOS — примерно 65 Гбайт:
  ```
  Filesystem     Size   Used  Avail Capacity iused      ifree %iused  Mounted on
  /dev/disk1s1  380Gi  304Gi   65Gi    83% 7757271 3974783129    0%   /System/Volumes/Data
  ```

Странно, в общем. Если проблема в ошибке, то можно запустить большое количество тестов локально, надеясь поймать её у себя.

Mazdaywik committed:
Почти рефакторинг: back-end не использует объявления функций (#315)

Почти рефакторинг, т.к. правка влияет на вычисления хешей областей видимости.
Но фактически это совершенно нейтральное изменение.

Mazdaywik committed:
Новые узлы дерева Externs, Drives, Inlines, Intrinsics (#315)

Формально это не рефакторинг, т.к. в режиме -OS может измениться порядок
пустых функций F@0, созданных для $EXTERN’ов, также изменился вывод
дерева в лог. Но фактически правки нейтральные.

Mazdaywik committed:
Стилевая правка: встроенную функцию объявлять не нужно

Mazdaywik committed:
Специализатор игнорирует шаблон специализации в дереве (#314, #359)

• На дно истории кладётся ГСО от образцов функции.
• Шаблон специализации удалён из всех форматов за ненадобностью.

Mazdaywik committed:
Почти рефакторинг: устранено дублирование кода (#314, #359)

Почти рефакторинг, т.к. в HighLevelRASL-ConjointFunc.ref переменные
получали индексы v.newNN, а теперь просто v.NN. Так что содержимое
лога изменилось. Но изменение нейтральное.

Mazdaywik committed:
Небольшой рефакторинг в OptTree-Spec.ref (#314, #359)

Mazdaywik committed:
Шаблон специализации удалён из рассахаренного дерева (#314, #359)

Почти рефакторинг, т.к. изменился вывод в отладочный лог.

В этом коммите OptTree-AutoMarkup-Spec тем не менее вычисляет шаблоны
специализации, но в дерево они не добавляются. Смысл в том, чтобы
сделать этот коммит почти рефакторингом — OptTree-AutoMarkup-Spec
должен помечать для специализации те же функции, что и раньше.

P.S. Этот коммит перезаписывает коммит ca86b6bb031, ошибочно приписанный
к заявке #315.

Mazdaywik committed:
FIXED: квадратичный размер дампа от размера поля зрения (#359, #361)

В дампе поля зрения вложенные вызовы печатаются с отступом, поэтому
в случае нехвостовой рекурсии размер дампа квадратично зависел от глубины
вложенности (отступы росли «треугольником»).

В случае, если в такой рекурсии происходило зацикливание, объём лога
мог быть неадекватно большим, что приводило к проблемам на сервере CI.

Теперь уровень вложенности ограничен 80 колонками, так что объём дампа
имеет линейную сложность от глубины вложенности.

Mazdaywik committed:
Улучшено форматирование списков имён в логе (#314)

Mazdaywik committed:
FIXED: специализируем только функции с хотя бы одним предложением (#314)

Mazdaywik committed:
Запрещаем в логе переменные без индекса (#315)

Mazdaywik committed:
FIXED: имена переменных при динамическом обобщении (#359)

Проблема была в том, что присваивания в динамическом обобщении теперь
последовательные, поэтому нужно тщательно выбирать имена переменных,
чтобы не зациклиться.

Mazdaywik committed:
Рефакторинги и стилевые правки (#359)

Mazdaywik committed:
Рефакторинг: уточнено имя автотеста (#160)

Mazdaywik committed:
FIXED: имена переменных при динамическом обобщении (#359)

Проблема была в том, что присваивания в динамическом обобщении теперь
последовательные, поэтому нужно тщательно выбирать имена переменных,
чтобы не зациклиться.

Коммит d7e729b010945 данную ошибку исправлял не полностью. Нужно было
таскать используемые имена вместе с обобщённым аргументом и e.Sg.

Mazdaywik committed:
Тюнинг настроек автотестов (#359)

• Подрегулированы пределы числа шагов и объёма памяти. Теперь они имеют
  минимально возможное значение, при котором выполняются тесты,
  вида 100…0, 200…0 или 500…0.
• Для некоторых тестов указан файл дампа. Файл дампа формируется быстрее,
  чем перенаправленный stderr, во всяком случае на Windows.

Mazdaywik committed:
Уточнение лимитов для тестового окружения + доработка CI (#359)

В артефактах теперь сохраняются отладочные дампы, добавленные предыдущим
коммитом.

Mazdaywik committed:
Небольшая оптимизация Solve-Clashes (#359)

Mazdaywik committed:
Повышение тестового лимита для GCC на CI (#359)

Mazdaywik committed:
Объём потребляемой памяти не зависит от размера префикса (#359)

Соответственно, снижены лимиты для тестов.

Mazdaywik committed:
FIXED: клэши-термы для экранируемых переменных (#359)

Mazdaywik committed:
Мелкие рефакторинги и стилевые правки (#359)

Mazdaywik committed:
Оптимизация перебора вариантов для клэшей e.X E : Pt P (#322, #314, #359)

Пояснения — см. в комментарии к исходному тексту.

Mazdaywik committed:
Усечение экранируемых предложений при специализации (#251, #314, #359)

Mazdaywik committed:
Опция -OA маркирует ВСЕ функции как специализируемые (#314)

Выяснилось, что на тесте opt-pattern5.sref началось нелинейное
(квадратичное) потребление памяти, из-за чего компилятор стал
вылетать по лимитам. Два предыдущих коммита были попытками обуздать
эту нелинейность, частично успешными. Нелинейность была побеждена
путём принудительного обобщения аргумента функции OutlineConstants
этого теста.

Ранее проблема не выявлялась, т.к. функции в тесте opt-pattern5.sref
не маркировались автоматически для специализации.

Для читабельности коммита в OptTree-AutoMarkup-Spec.ref был добавлен
новый код без удаления старого. Старый будет удалён в следующем
коммите.

Обнаружен случайный тест, требующий более 100 000 000 шагов (конкретно,
за 111 622 480 в режиме -OADS). Тест был добавлен как сохранённый.
В дальнейшем, если удастся оптимизировать компилятор и тест будет
проходить быстрее, порог можно поднять, а тест можно будет удалить.

Mazdaywik committed:
Удалён старый код разметки специализируемых функций (#314)

Mazdaywik committed:
Предупреждения для объявлений $INTRINSIC для определённых функций (#314)

Объявления $INTRINSIC игнорируются для функций с определениями. Об этом
выводятся предупреждения, также себя ведёт и рассахариватель (удаляет
ненужные пометки).

OptTree-Drive.ref учитывает тот факт, что множества интринсиков
и прогоняемых функций не пересекаются.

Mazdaywik committed:
Стилевые правки: оформление кода (#314, #359)

Коммит смотреть с опцией игнорирования пробелов.

Mazdaywik committed:
Прогонщик считает директиву $INLINE синонимом для $DRIVE (#314)

Изменения внесены на уровне прогонщика, т.к. для целей отладки
остальные компоненты (в частности, разметка для прогонки) оставлены
прежними. Разметка для прогонки по-разному интерпретирует метки
$INLINE и $DRIVE.

Любопытно, что этот коммит замедляет время работы компилятора в режиме
-OiDRPS. Был сделан замер с 9 проходами с параметрами

set BENCH_FLAGS=-OiDPRS
set RLMAKE_FLAGS=-X-OiDPRS

до этого коммита и на нём.

• До:   (Total refal time): 125,20100 [124,787…125,691],
        Step count 175 608 086,
• После (Total refal time): 132,17600 [132,025…132,672],
        Step count 191 734 143.

Число шагов возросло на 9,2 %, время работы — на 5,6 %. Операций
стало выполняться больше, но производительность при этом
возросла.

Mazdaywik committed:
FIXED: прогонка вызова метатаблицы рушила компилятор (#314, #359)

Mazdaywik committed:
Имена метатаблиц не распечатываются в логе (#314)

Mazdaywik committed:
Рефакторинг структур данных алгоритма прогонки (#314)

Mazdaywik committed:
Опция -OA маркирует ВСЕ функции как прогоняемые (#314)

Для читабельности коммита в OptTree-AutoMarkup-Drive.ref был добавлен
новый код без удаления старого. Старый будет удалён в следующем
коммите.

В тесте opt-tree-spec12.ref заблокирована возможность прогонки
и специализации функции F, т.к. с новым алгоритмом разметки он перестал
входит в лимит числа шагов.

Mazdaywik committed:
Компилятор самоприменяется в режиме -X-OiADPRS -X--log=… (#314, #332)

Предыдущий коммит опять углубил возможности оптимизации, из-за чего
пришлось выявлять и купировать очередные распухания.

Mazdaywik committed:
FIXED: функция gen_e__ также не специализируется (#331, #332)

Проблема возникала на этапе прогонки экземпляров: на первой стадии
оптимизации функция gen_e__ оберегала от взрывной прогонки,
но на второй фазе получаются прогоняемые экземпляры gen_e__@n,
которые уже ничего не защищают. И программа распухает на финальной
стадии.

Mazdaywik committed:
Переименование автотестов на отношение Хигмана-Крускала (#359)

Отношение Хигмана-Крускала теперь проверяется не только для
специализации, но и для прогонки, поэтому тесты имело смысл
переименовать.

Mazdaywik committed:
FIXED: отношение остановки не учитывает «замороженность» скобок (#359)

Ранее из-за этого проверка вложения давала отрицательный результат,
если опасно похожие выражения различались замороженностью. Из-за этого
иногда разворачивался лишний виток цикла.

Автотест специально написан так, чтобы в старой версии программа
чрезмерно распухала и тест вылетал с лимитом шагов.

Mazdaywik committed:
Мелкие рефакторинги (#359)

Mazdaywik committed:
FIXED: экранирование при прогонке, откат opt-tree-spec12.ref (#314, #359)

Добавлено распознавание экранирования, как описано в комментарии
https://github.com/bmstu-iu9/refal-5-lambda/issues/359#issuecomment-902277608

Экранирование распознаётся только в самом низу, при ветвлении. На это
есть две причины:

1. Если распознавание экранирования откладывать, то «ациклическая
   суперкомпиляция» продолжит прогонять бесполезные ветви. Это лишние
   шаги и лишнее увеличение дерева (которое может достичь и лимита
   узлов).
2. Формально этого достаточно для прохождения теста opt-tree-spec12.ref
   в исходном виде, поэтому дополнительно усложнять остальной код
   я не стал. Экранирования между разными ветвями некритичны.

Mazdaywik committed:
FIXED: раскраска Простого Рефала для Far Manager поддерживает $INTRINSIC

Mazdaywik committed:
Revert "FIXED: раскраска Простого Рефала для Far Manager поддерживает $INTRINSIC"

This reverts commit 01124a98df7fafd0d66e13e580e312ebadedbb1b.

Mazdaywik committed:
Перевод opt-pattern5.sref на Рефал-5λ: 1 — просто переименование (#359)

Mazdaywik committed:
Перевод opt-pattern5.sref на Рефал-5λ: 2 — обновление синтаксиса (#359)

Mazdaywik committed:
FIXED: распухание кода в автотесте opt-pattern5.ref (#332, #359)

Распухание происходило из-за прогонки функции Inc. Решение: заменить
это определение функции на пометку $INTRINSIC.

Но Простой Рефал не поддерживает ключевое слово $INTRINSIC, поэтому
автотест пришлось сначала перевести на Рефал-5λ двумя предыдущими
коммитами.

Mazdaywik committed:
Небольшая оптимизация Solve-SymmClashes (#359)

Коммит вводит «таксономию» для симметричных клэшей аналогично
коммиту 629f85b8c.

Mazdaywik committed:
Оптимизация сопоставления с неразменными переменными (#231, #314, #359)

Оптимизация была необходима, т.к. перестал проходить тест
saved-test-77_14.08.2021_20-00-01,67-big.ref (см. #314, 41ce59e8d0).

Теперь снова проходит.

Mazdaywik committed:
Оптимизация симметричного клэша с замыканием (#322, #359)

Mazdaywik committed:
Уточнения имён переменных (#359)

Mazdaywik committed:
Уточнение прелюдии для случайных тестов (#314)

Со старой версией выдавались предупреждения согласно
https://github.com/bmstu-iu9/refal-5-lambda/issues/314#issuecomment-899034440

То, что функции теперь помечены $EXTERN — не страшно. Всё равно
эти файлы не линкуются, поэтому определения встроенным функциям не нужны.

Mazdaywik committed:
Чистка OptTree-AutoMarkup-Drive.ref от лишнего кода (#314, #359)

Mazdaywik committed:
Оптимизация применения сужений к выражению (#359)

Mazdaywik committed:
Принудительные обобщения в куче мест для борьбы с распуханием № 5 (#359)

По-нормальному, нужно каждый случай обобщения описать, сделать выводы
и решить проблему в общем случае в рамках #332. А пока я анализировал
лог и подсекал наиболее явные точки распухания.

Mazdaywik committed:
Уточнён принцип нумерации новых переменных

Mazdaywik committed:
Устаревшие узлы дерева переименованы в Legacy… (#314, #318)

Mazdaywik committed:
Псевдокомментарии \*\$(NO-)?(OPT|DRIVE|SPEC) (#195, #314)

После псевдокомментария может следовать любое количество имён, запятых
и точек с запятой (знаки пунктуации игнорируются), имя должно быть
как минимум одно.

Mazdaywik committed:
Использование *$NOOPT вместо gen_e__ в некоторых местах (#332, #359)

Mazdaywik committed:
Рефакторинг: удалён неактуальный кусок кода (#359)

Код стал неактуальным после коммита 855d138e8c, изменившего логику
присваиваний при динамическом обобщении (коммит был с ошибками,
его исправления d7e729b01 и ff803b5c19).

Mazdaywik committed:
Исправлены некоторые TODO

Mazdaywik committed:
В автотестах на оптимизацию используются псевдокомментарии (#195, #314)

Mazdaywik committed:
FIXED: не работали -Wall -Werror для pseudocomments и intrinsic (#314)

Mazdaywik committed:
FIXED: псевдокомментарии *$OPT и нативные функции (#314)

В этом случае выдаётся предупреждение, а псевдокомментарий игнорируется.

Если функция объявлена как $SWAP, то предупреждений не выдаётся, пометка
просто молча игнорируется.

Для псевдокомментариев $NO… такой проверки не делается, т.к. это
не имеет смысла.

Mazdaywik committed:
Рассахариватель считает ключевое слово $INLINE синонимом $DRIVE (#314)

В коммите 18e35b7c7c7c9 их не различал только прогонщик, поскольку
из соображений отладки сохранялось поведение авторазметки (меньше
изменений поведения в одном коммите). Теперь, когда авторазметка
упрощена, их можно не различать и на более ранней стадии. Смешивать
их на стадии синтаксического анализа пока нельзя, т.к. их должна
различать проверка семантики (выдавать адекватные сообщения об ошибках).

Mazdaywik committed:
-Wpseudocomments как синоним для -Wpseudocomment (#195, #314)

Mazdaywik committed:
Стилевая правка

Mazdaywik committed:
FIXED: запрет оптимизации для gen-функций (#314, #331)

Во-первых, если псевдокомментарий разрешает оптимизацию и включает
имя gen-функции, то выдаётся предупреждение (даже если сама
gen-функция написана неправильно). Ранее эта функция не работала,
т.к. в коде CheckPseudofunction была ошибка.

Во-вторых, поскольку предупреждение не блокирует дальнейшую
компиляцию, на стадии рассахаривания для gen-функций оптимизация
запрещается (имена помещаются в узлы (NoDrives …) и (NoSpecs …)).

Поскольку оптимизация gen-функций заранее отсекается на стадии
рассахаривания, в авторазметках можно их не учитывать.

Mazdaywik committed:
Рефакторинг: устранение дублирования имён gen-функций (#314, #331)

Mazdaywik committed:
Переим. OptTree-AutoMarkup-Drive.ref → OptTree-AutoMarkup.ref (#359)

Слияние выполняется через два последовательных переименования
для сохранения истории правок.

Mazdaywik committed:
Переим. OptTree-AutoMarkup-Spec.ref → OptTree-AutoMarkup.ref (#359)

Слияние выполняется через два последовательных переименования
для сохранения истории правок.

Mazdaywik committed:
Слияние двух файлов в файл OptTree-AutoMarkup.ref (#359)

Слияние выполняется через два последовательных переименования
для сохранения истории правок. История правок сохранилась,
даже сохранились (в выводе git blame) некоторые строчки
авторства @Kaelena!

Mazdaywik committed:
Удалён неактуальный проход «Cure drives» (#314, #359)

После удаления интеллектуальной авторазметки данный проход стал пустым,
теперь он и вовсе удалён. Это не рефакторинг, поскольку содержимое
лога изменилось.

Mazdaywik committed:
Уточнение документации на синтаксическое дерево (#314)

Mazdaywik committed:
Псевдокомментарий *$IDENT как замена ключевого слова $LABEL (#195)

В заявке #195 предлагалось заменить $LABEL на *$PRAGMA-NATIVE-IDENTS,
но было выбрано слово $IDENT. Во-первых, вещь второстепенная
и с удалением нативных вставок (см. #318) она будет удалена, во-вторых,
имеющийся код легко расширить новым ключевым словом, если оно совпадает
с точностью до регистра с узлом дерева. А директива $LABEL отображается
на узел дерева (Ident t.Pos e.Name).

Mazdaywik committed:
src/lib: псевдокомментарии вместо ключевых слов (#195, #314)

Mazdaywik committed:
src/compiler: псевдокомментарии вместо ключевых слов (#195, #314)

Mazdaywik committed:
Выдача -Wdeprecated на $DRIVE, $INLINE, $LABEL, $SPEC (#195,#314,#318)

Mazdaywik committed:
FIXED: совместимость со стабильной версией (#195, #314)

Стабильная версия иначе интерпретировала псевдокомментарии и выдавала
на них ошибки компиляции. Для обеспечения совместимости внедрён костыль
в виде псевдокомментариев вида *$xDRIVE. После обновления стабильной
версии данный костыль нужно будет удалить.

Подробности в комментарии в коде.

Mazdaywik committed:
Revert "Исправлены некоторые TODO"

This reverts commit 2b6be6371315a9e92be5d1a70b0c260f54ca1509.

Одно TODO было «исправлено» без понимания проблемы, его откатил.

Mazdaywik committed:
Замена gen_e__ на *$NODRIVE в OptTree-Spec.ref (#332, #359)

Mazdaywik committed:
Мелкие рефакторинги (#362)

• Устранение дублирования кода.
• Прояснение имён переменных.

Mazdaywik committed:
Отдельный случай простого зацикливания (#362)

Рефакторингом в чистом виде он, скорее всего, не является,
т.к. может изменить имена переменных в экземплярах, наблюдаемые в логе.

Mazdaywik committed:
Рефакторинг для следующего коммита (#362)

Mazdaywik committed:
История строится как при перестройке сверху (#362)

Содержимое лога для теста bloat-basic.ref (приложенного к заявке)
не изменилось за исключением вывода самой истории. Т.е. на данный
тест замена никак не повлияла.

На других примерах работу не испытывал.

Mazdaywik committed:
Рефакторинг: ручная специализация по s.NeedRelationCheck (#362)

Фактически, развёрнут последний виток цикла.

Mazdaywik committed:
Рефакторинг: очевидные упрощения кода (#362)

Mazdaywik committed:
Рефакторинг цепочек вызовов функций (#362)

• Удалено промежуточное звено SpecCall-BuildSignature-***,
• функция SpecCall-CheckSignature выпала из цепочки и стала
  нерекурсивной (избыточные параметры удалены).

Mazdaywik committed:
Шаг к перестройке сверху — функции-патчи (#362)

Mazdaywik committed:
Рефакторинг кода специализации с обобщениями (#362)

Mazdaywik committed:
Реализована «халтурная» перестройка сверху (#362)

Экземпляры строятся как обычно, но при обнаружении зацикливания
более ранний экземпляр в истории также подменяется вызовом обобщённого
вроде

  Loop@3 {
    (e.0) t.1 t.2 e.3 = <Loop@8 (e.0) t.1 W S t.2 e.3>;
  }

Mazdaywik committed:
Нумерация переменных после обобщения начинается с 0 (#362)

Перемена нумерации позволяет тривиальным сигнатурам на дне истории
(сигнатурам псевдоэкземпляров Func@0) выглядеть в логеконсистентно
с последующими сигнатурамии Func@n, где нумерация начинается с 0.

Mazdaywik committed:
Автотесты для #362 (НЕ ПРОХОДЯТ!!!) (#362)

Два автотеста недописаны (нет функции Go) и поэтому выключены
(переименованы в *._ref).

Случайно сгенерированный автотест, заваливший CI, сохранён и на нём
компилятор по-прежнему падает.

Смысл коммита — не потерять эти тесты. Когда проблема #362 будет
исправлена, тесты opt-tree-bloat-*._ref будут дописаны и включены.

Mazdaywik committed:
Функции OptTree-***-Prepare забывают предыдущую разметку (#362)

Для специализации это очевидный рефакторинг, т.к. разметка выполняется
не более чем один раз в любом режиме компиляции. (Вообще, этот код
давно надо было упростить.)

Для прогонки это не рефакторинг, т.к. при прогонке экземпляров
содержимое лога будет немного другим.

Помимо непосредственного удаления кода «обновления» структур данных,
были сделаны некоторые простые рефакторинги в окружающем коде.

Mazdaywik committed:
Построение let-экземпляров (#362)

Mazdaywik committed:
Пропатченные и let-экземпляры теперь встраиваются и чистятся (#362)

Mazdaywik committed:
Лимит на предельное число экземпляров повышен с 100 до 150 (#332, #362)

Ранее было обнаружено, что с новым алгоритмом специализации компилятор
в некоторых тестах увязает, и был добавлен грязный хак, ограничивающий
максимальное число экземпляров (54c28cad, #332).

Предыдущие два коммита стали строить дополнительные let-экземпляры, из-за
чего порог стал достигаться раньше и глубина оптимизации снизилась.
Поэтому потребовалось поднять порог.

Порог был поднят до 150 по двум причинам:

• В тесте saved-test-10_Mon-Aug-23-21-51-16-UTC-2021.ref (81667ba)
  на пороге в 100 экземпляров условие остановки не срабатывает
  (до let-экземпляров срабатывало). Опытным путём было выяснено,
  что условие установки срабатывает на пороге между 140 и 145, было
  выбрано круглое значение.

  Заметим, что этот тест в текущем коммите проходит.

• Тест saved-test-77_14.08.2021_20-00-01,67-big.ref (41ce59e8, #314),
  напротив, проходит только благодаря наличию этого лимита (на момент
  написания коммита 41ce59e8 не анализировалась проблема
  с непрохождением теста, просто был повышен порог). С лимитом 200
  тест не проходит, с лимитом 150 проходит.

Mazdaywik committed:
Рефакторинг: функции перемещены ближе к точкам вызовов (#362)

Mazdaywik committed:
FIXED: проход прогонки экземпляров прогоняет только экземпляры (#362)

Конкретная проблема предыдущего коммита была в том, что на вход финальной
прогонки экземпляров подавалось дерево с функцией Func*n, но при этом
функция Func в дереве могла отсутствовать. Для прогонщика это было
нарушением инварианта.

Исправление свелось к тому, что Func*n уже больше не подлежат финальной
прогонке, т.к. подлежат только экземпляры — функции с другим суффиксом @n.

Была и другая проблема (не проявлялась). Она была в том, что ранее
на финальном этапе выполнялась разметка всех функций для прогонки,
а функции с именами вида Func*n прогонялись уже не адекватно,
т.к. предложения в них уже могли быть размножены, и прогонка одного
предложения приводила бы к неправильной функции Func*‹n+1›.

Mazdaywik committed:
Мелкие рефакторинги и стилевые правки (#362)

Mazdaywik committed:
Некоторый рефакторинг прогонки выражений (#230, #362)

Mazdaywik committed:
Поддержка прогонки вызовов с активными аргументами (#230, #362)

Обнаружилась проблема, связанная с исчерпанием числа шагов на тесте
varcopy-fail.ref. Я предположил, что проблема в лишних let-экземплярах,
которые съедают один шаг. Поэтому всё-таки реализовал #230 в рамках
исправления #362.

Как оказалось, реализовать поддержку активных вызовов оказалось проще,
чем я думал ранее.

Mazdaywik committed:
FIXED: частный случай обобщения при имитации перестройки сверху (#362)

Проблему прекрасно описывает обширный комментарий в коде.

Тест varcopy-fail.ref переименован, т.к. он выявлял проблему, решённую
в коде.

Mazdaywik committed:
FIXED: подстановки при прогонке вызовов с активными аргументами (#230)

Mazdaywik committed:
FIXED: уточнён предохранитель распухания (#332, #362)

Теперь специализация одного вызова может порождать несколько экземпляров,
а значит, сравнение на точное значение может «перескочить».

TonitaN reviewed (dismissed):
Совершенству нет предела. Скоро новый релиз?

Mazdaywik commented:
> Совершенству нет предела. Скоро новый релиз?

Когда появится коммит, меняющий [этот файл](https://github.com/bmstu-iu9/refal-5-lambda/blob/master/NEWS.md).

Но вообще, в конце комментария https://github.com/bmstu-iu9/refal-5-lambda/issues/362#issuecomment-929874727 поставлен вопрос, на который я бы хотел сначала ответить. Самому интересно, на сколько актуален сейчас этот проход. А так да, совершенству нет предела.

Mazdaywik committed:
Стилевая правка: переводы строк на концах исходников

Mazdaywik committed:
FIXED2: подстановки при прогонке вызовов с активными аргументами (#230)

Предыдущий коммит d5847ba94f устранял падение компилятора, скрывая ошибку:
оптимизация просто переставала работать.

Mazdaywik committed:
Merge branch 'master' into v3.4

Mazdaywik committed:
FIXED: лимит памяти для тестов autotests/Library повышен (#341)

После вливания правок по #341 в FindErrors в Checker.ref увеличилась
потребность в памяти — дерево e.AST теперь копируется:

    <CheckScreening (e.WarningIds) (e.AST)>
    <CheckMultiScreening (e.WarningIds) (e.AST)>;

Поэтому пришлось повысить лимит для компиляции Library.ref
в соответствующей папке.

Mazdaywik committed:
FIXED: актуальная версия Far Colorer требует кодировку UTF-8

Mazdaywik committed:
Merge branch 'master' into v3.4

Mazdaywik committed:
Повышение точности динамического обобщения (#251, #322, #362)

Mazdaywik committed:
Рефакторинг: режим переменной mode, уточнение комментариев

Mazdaywik committed:
Log-InlineExpr можно использовать при отладке GenericMatch

Mazdaywik committed:
Отладочная печать сигнатур специализируемых функций

Mazdaywik committed:
Повышен лимит шагов для раскрутки в режиме -OiADPRS

Mazdaywik committed:
Merge branch 'master' into v3.4
